include/mbedtls/constant_time.h

```diff
@@ -9,6 +9,10 @@
 #ifndef MBEDTLS_CONSTANT_TIME_H
 #define MBEDTLS_CONSTANT_TIME_H
 
```

Comment:
We currently validate C++ compatibility only with by compiling a [trivial program that includes all headers but does nothing](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.6.0/programs/test/generate_cpp_dummy_build.sh). This rejects, for example, `extern "C" {` without the matching `}`, but doesn't detect if it's misplaced or missing. Can we do better and catch a missing or misplaced `extern "C"`, without going through the trouble of calling every function?

Reply:
An idea would be to add a lint to CI that:
- Has a list of "important" headers that actually declare functions. I think there are some macro-only header-files, which don't need extern C.
- Greps each file in the list for the exact multi-line strings
```c
#ifdef __cplusplus
extern "C" {
#endif
```
I don't know how the linting infrastructure works here. Are there any scripts you could point me to?

Reply:
These are the errors generated currently for ref. We'd need some more elegant way to mark the headers that are to be excluded.
<spoiler>
nikita@fedora-3:~/src/mbedtls$ ./programs/test/generate_cpp_dummy_build.sh 
Error: include/mbedtls/build_info.h does not contain a __cplusplus guard
Error: include/mbedtls/check_config.h does not contain a __cplusplus guard
Error: include/mbedtls/compat-2.x.h does not contain a __cplusplus guard
Error: include/mbedtls/private_access.h does not contain a __cplusplus guard
Error: include/mbedtls/psa_util.h does not contain a __cplusplus guard
Error: include/psa/build_info.h does not contain a __cplusplus guard
Error: include/psa/crypto_adjust_auto_enabled.h does not contain a __cplusplus guard
Error: include/psa/crypto_adjust_config_key_pair_types.h does not contain a __cplusplus guard
Error: include/psa/crypto_adjust_config_synonyms.h does not contain a __cplusplus guard
Error: include/psa/crypto_builtin_composites.h does not contain a __cplusplus guard
Error: include/psa/crypto_builtin_key_derivation.h does not contain a __cplusplus guard
Error: include/psa/crypto_builtin_primitives.h does not contain a __cplusplus guard
Error: include/psa/crypto_driver_common.h does not contain a __cplusplus guard
Error: include/psa/crypto_driver_contexts_composites.h does not contain a __cplusplus guard
Error: include/psa/crypto_driver_contexts_key_derivation.h does not contain a __cplusplus guard
Error: include/psa/crypto_driver_contexts_primitives.h does not contain a __cplusplus guard
Error: include/psa/crypto_legacy.h does not contain a __cplusplus guard
Error: include/psa/crypto_platform.h does not contain a __cplusplus guard
Error: include/psa/crypto_sizes.h does not contain a __cplusplus guard
Error: include/psa/crypto_types.h does not contain a __cplusplus guard
Error: include/psa/crypto_values.h does not contain a __cplusplus guard
Some headers are missing __cplusplus guards
</spoiler>

Reply:
Thanks for checking! The only testing we're currently doing is compiling [a generated program](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.6.0/programs/test/generate_cpp_dummy_build.sh) in one configuration. (Testing [more configurations](https://github.com/Mbed-TLS/mbedtls/issues/8331) would be good, but out of scope here.)

Several headers are skipped because they aren't meant to be included directory. In addition, some headers only declare macros (`*/build_info.h`, `*/private_access.h`). That leaves `mbedtls/compat-2.x.h` (can't be very important since nobody's complained) and `mbedtls/psa_util.h` (a recent addition). Would you mind adding guards to those as well while you're at it?

Reply:
I won't ask for a non-regression test because this is a long-standing issue and we don't add headers often. But if you're willing to dive into it, linting is done from the `component_check_*` functions in [`tests/scripts/all.sh`](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.6.0/tests/scripts/all.sh).